src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs

```diff
@@ -26,37 +27,137 @@ use itertools::Itertools;
     smallest_table_key: Vec<u8>,
     size: usize,
     _tracker: Option<MemoryTracker>,
+    /// For a batch created from multiple batches, this will be
+    /// the largest batch id among input batches
     batch_id: SharedBufferBatchId,
 }
 
 impl SharedBufferBatchInner {
-    fn new(
+    pub(crate) fn new(
+        epoch: HummockEpoch,
         payload: Vec<SharedBufferItem>,
-        mut range_tombstone_list: Vec<DeleteRangeTombstone>,
+        range_tombstone_list: Vec<DeleteRangeTombstone>,
         size: usize,
         _tracker: Option<MemoryTracker>,
     ) -> Self {
+        let (smallest_empty, mut smallest_table_key, mut largest_table_key, range_tombstones) =
+            Self::get_table_key_ends(range_tombstone_list);
+
+        if let Some(item) = payload.last() {
+            if item.0.gt(&largest_table_key) {
+                largest_table_key.clear();
+                largest_table_key.extend_from_slice(item.0.as_ref());
+            }
+        }
+        if let Some(item) = payload.first() {
+            if smallest_empty || item.0.lt(&smallest_table_key) {
+                smallest_table_key.clear();
+                smallest_table_key.extend_from_slice(item.0.as_ref());
+            }
+        }
+        let items = payload
+            .into_iter()
+            .map(|(k, v)| (k, vec![(epoch, v)]))
+            .collect_vec();
+
+        let batch_id = SHARED_BUFFER_BATCH_ID_GENERATOR.fetch_add(1, Relaxed);
+        SharedBufferBatchInner {
+            payload: items,
+            epoch,
+            has_multi_epochs: false,
+            imm_ids: vec![batch_id],
+            epochs: vec![epoch],
+            range_tombstone_list: range_tombstones,
+            size,
+            largest_table_key,
+            smallest_table_key,
+            _tracker,
+            batch_id,
+        }
+    }
+
+    pub(crate) fn new_with_multi_epoch_batches(
+        min_epoch: HummockEpoch,
+        epochs: Vec<HummockEpoch>,
+        payload: Vec<SharedBufferVersionedEntry>,
+        imm_ids: Vec<ImmId>,
+        range_tombstone_list: Vec<DeleteRangeTombstone>,
+        size: usize,
+        tracker: Option<MemoryTracker>,
+    ) -> Self {
+        debug_assert!(!imm_ids.is_empty());
+
+        let (smallest_empty, mut smallest_table_key, mut largest_table_key, range_tombstones) =
+            Self::get_table_key_ends(range_tombstone_list);
+
+        if let Some(item) = payload.last() {
+            if item.0.gt(&largest_table_key) {
+                largest_table_key.clear();
+                largest_table_key.extend_from_slice(item.0.as_ref());
+            }
+        }
+        if let Some(item) = payload.first() {
+            if smallest_empty || item.0.lt(&smallest_table_key) {
+                smallest_table_key.clear();
+                smallest_table_key.extend_from_slice(item.0.as_ref());
+            }
+        }
+
+        let max_imm_id = *imm_ids.iter().max().unwrap_or(imm_ids.first().unwrap());
```

Comment:
shouldn't `imm_ids.iter().max().unwrap()` be sufficient since `imm_ids` cannot be empty?